### PR TITLE
[SPARK-23197][STREAMING][TESTS] Fix ReceiverSuite."receiver_life_cycle" to not rely on timing

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This patch changes ReceiverSuite."receiver_life_cycle" to record actual calls with timestamp in FakeReceiver/FakeReceiverSupervisor, which doesn't rely on timing of stopping and starting receiver in restarting receiver. It enables us to give enough huge timeout on verification of restart as we can verify both stopping and starting together.

### Why are the changes needed?

The test is flaky without this patch. We increased timeout to fix flakyness of this test (https://github.com/apache/spark/commit/15adcc8273e73352e5e1c3fc9915c0b004ec4836) but even with longer timeout it has been still failing intermittently.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

I've reproduced test failure artificially via below diff:

```
diff --git a/streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala b/streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala
index faf6db82d5..d8977543c0 100644
--- a/streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala
+++ b/streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala
@@ -191,9 +191,11 @@ private[streaming] abstract class ReceiverSupervisor(
       // thread pool.
       logWarning("Restarting receiver with delay " + delay + " ms: " + message,
         error.getOrElse(null))
+      Thread.sleep(1000)
       stopReceiver("Restarting receiver with delay " + delay + "ms: " + message, error)
       logDebug("Sleeping for " + delay)
       Thread.sleep(delay)
+      Thread.sleep(1000)
       logInfo("Starting receiver again")
       startReceiver()
       logInfo("Receiver started again")
```

and confirmed this patch doesn't fail with the change.

### CODE REVIEW

File: streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala [hunk 1]
```diff
@@ -106,19 +106,22 @@ class ReceiverSuite extends TestSuiteBase with TimeLimits with Serializable {
...
-      assert(receiver.onStopCalled)
-    }
-    eventually(timeout(1.second), interval(10.milliseconds)) {
-      // receiver will be started async
-      assert(receiver.onStartCalled)
-      assert(executor.isReceiverStarted)
+    executor.callsRecorder.reset()
+    receiver.callsRecorder.reset()
+    receiver.restart("restarting", null, 100)
+    eventually(timeout(10.seconds), interval(10.milliseconds)) {
```
Comment (dongjoon-hyun) [inline streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: So, `10.seconds` is enough? Do you need to retriever this PR to validate more?
BTW, thank you so much for taking care of this case!

Comment (HeartSaVioR) [reply streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: Yes that was actually 1.3 seconds (300ms + 1s) and it hasn't been failing for high probability so it should be pretty enough.

**Comment by HeartSaVioR** [comment]
cc. @tdas @sameeragarwal as they've authored and reviewed 5adcc8
also cc. to @zsxwing @jose-torres @gaborgsomogyi

File: streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala [hunk 2]
```diff
@@ -106,19 +106,22 @@ class ReceiverSuite extends TestSuiteBase with TimeLimits with Serializable {
...
-      assert(receiver.isStopped)
-      assert(receiver.onStopCalled)
-    }
-    eventually(timeout(1.second), interval(10.milliseconds)) {
-      // receiver will be started async
-      assert(receiver.onStartCalled)
-      assert(executor.isReceiverStarted)
+    executor.callsRecorder.reset()
+    receiver.callsRecorder.reset()
+    receiver.restart("restarting", null, 100)
```
Comment (dongjoon-hyun) [inline streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: This goes down from 600 to 100?

**Review by dongjoon-hyun — APPROVED** [review]
I also verified this PR with the given test method. New test suite is logically robust and doesn't fail.
+1, LGTM. Merged to master.

Thank you so much, @HeartSaVioR !

**Comment by dongjoon-hyun** [comment]
Could you make a backporting PR, @HeartSaVioR ?

File: streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala (hunk 2 above)
Comment (HeartSaVioR) [reply streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: Ah yes we no longer need so long delay as we don't rely on timing.

**Comment by HeartSaVioR** [comment]
Thanks for reviewing and merging, @dongjoon-hyun ! I'll submit a PR for branch-2.4 as well.

**Comment by HeartSaVioR** [comment]
#25930

**Comment by dongjoon-hyun** [comment]
Thanks!

**Pull request closed without merging**